scompose/tests/test_depends_on.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python
+
+# Copyright (C) 2019-2020 Vanessa Sochat.
+
+# This Source Code Form is subject to the terms of the
+# Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+# with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from scompose.project import Project
+from scompose.utils import run_command
+from time import sleep
+import shutil
+import pytest
+import os
+
+here = os.path.dirname(os.path.abspath(__file__))
+
+
+def test_circular_dependency(tmp_path):
+
+    depends_on = os.path.join(here, "configs", "wrong_depends_on")
+    for filename in os.listdir(depends_on):
+        source = os.path.join(depends_on, filename)
+        dest = os.path.join(tmp_path, filename)
+        print("Copying %s to %s" % (filename, dest))
+        shutil.copyfile(source, dest)
+
+    # Test the simple apache example
+    os.chdir(tmp_path)
+
+    # Check for required files
+    assert "singularity-compose.yml" in os.listdir()
+
+    print("Creating project...")
+
+    # Loading project validates config
+    project = Project()
+
+    print("Testing build")
+    project.build()
+
+    for image in ["first.sif", "second.sif", "third.sif"]:
+        assert image in os.listdir(tmp_path)
+
+    print("Testing view config")
+    project.view_config()
+
+    try:
+        print("Testing up")
+        project.up()
+        raise Exception("Up should have failed")
+    except SystemExit:
+        print("Up failed as expected")
+    finally:
+        print("Bringing down")
+        project.down()
+
+
+def test_no_circular_dependency(tmp_path):
+    depends_on = os.path.join(here, "configs", "depends_on")
+    for filename in os.listdir(depends_on):
+        source = os.path.join(depends_on, filename)
+        dest = os.path.join(tmp_path, filename)
+        print("Copying %s to %s" % (filename, dest))
+        shutil.copyfile(source, dest)
+
+    # Test the simple apache example
+    os.chdir(tmp_path)
+
+    # Check for required files
+    assert "singularity-compose.yml" in os.listdir()
+
+    print("Creating project...")
+
+    # Loading project validates config
+    project = Project()
+
+    print("Testing build")
+    project.build()
+
+    for image in ["first.sif", "second.sif", "third.sif"]:
+        assert image in os.listdir(tmp_path)
+
+    print("Testing view config")
+    project.view_config()
+
+    # Test depends_on DAG order
```

Comment:
The one addition I was thinking would make sense to test is to be able to have the instances echo / write to some shared file about the order they are starting up in (that's why the startscripts have the echos for now). What do you think?

Reply:
Oh ok, so your idea is to test the output rather than the instances list in case the internal code would change, right?

Reply:
I think we might do well to test the instances list (as you've done) and the actual output, in case the way we have them listed doesn't correspond to the start order.